pkg/controller/virtualmachineimport/virtualmachineimport_controller.go

```diff
@@ -232,6 +232,37 @@ func (r *ReconcileVirtualMachineImport) Reconcile(request reconcile.Request) (re
 		return reconcile.Result{}, err
 	}
 
+	// Add finalizer to handle cancelled import
+	if !r.vmImportInProgress(instance) {
+		err := utils.AddFinalizer(instance, utils.CancelledImportFinalizer, r.client)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Handle deleted import
+	if instance.DeletionTimestamp != nil {
```

Comment:
We already check for deletion timestamp. Please unify and extract a method.

Reply:
We check after we have a provider. This is for the only existing finalizer. However, this new finalizer does not rely on anything but the cluster itself. Therefore, I can not unify. I tried. I can do something very artificial and in my opinion not so pretty. I can create a single function that takes all possible arguments. Inside in case provider is nil do part A and if not nil do part B. I think this is not such a good approach.

Reply:
My suggestion was to unify cancellation logic. With your PR we are handling it in 2 places. I think we could have something like:

```go
// CR cancelled
if instance.DeletionTimestamp != nil {
 // handle both
}
```

Whether externalizing it or keeping in reconcile it is up to you. I think writing tests for this logic would help us understand which approach would be better.

Reply:
I will add tests, of course.
Regarding the handling of cancellation. It is just in this location.
I add finalizer when resource is first handled.
Remove it when it is done.
Call it if resource was deleted.
The other test for deletion is for fializer that requires provider.